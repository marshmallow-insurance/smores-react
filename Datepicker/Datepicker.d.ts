import { FC } from 'react';
import { MarginProps } from '../utils/space';
export type DatepickerProps = {
    showDayLabels?: boolean;
    disableWeekend?: boolean;
    fromDate?: Date;
    endingDate?: Date;
    showYear?: boolean;
    showSelectedDate?: boolean;
    range?: number;
    onDateSelect: (date: string) => void;
    onChange?: (value: Date) => void;
    value?: Date;
    fallbackStyle?: boolean;
} & MarginProps;
export declare const Datepicker: FC<DatepickerProps>;
