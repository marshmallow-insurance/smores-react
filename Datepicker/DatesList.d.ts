import { FC } from 'react';
import { Day } from './types';
type Props = {
    items: Day[];
    showDayLabels: boolean;
    handleDateSelect: (date: Date) => void;
};
export declare const DatesList: FC<Props>;
export {};
