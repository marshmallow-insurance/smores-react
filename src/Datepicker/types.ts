export interface Day {
  label: string
  disabled: boolean
  active: boolean
  date: Date
}
