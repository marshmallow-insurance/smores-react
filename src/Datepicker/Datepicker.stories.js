import React from 'react'
import { Datepicker } from './Datepicker'

export default {
  title: 'Datepicker',
  component: Datepicker,
  argTypes: { onDateSelect: { action: 'date selected' } },
}

const Template = (args) => <Datepicker {...args} />

export const Default = Template.bind({})

Default.args = {
  disableWeekend: false,
  firstDayShift: true,
  range: 31,
  fromDate: new Date('2021/07/14'),
}
