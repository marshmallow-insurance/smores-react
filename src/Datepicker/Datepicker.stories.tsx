import React from 'react'
import { Datepicker, DatepickerProps } from './Datepicker'

export default {
  title: 'Datepicker',
  component: Datepicker,
  argTypes: { onDateSelect: { action: 'date selected' } },
}

const Template = (props: DatepickerProps) => <Datepicker {...props} />

export const Default = Template.bind({})

Default.args = {
  showDayLabels: false,
  disableWeekend: false,
  firstDayShift: true,
  range: 31,
  fromDate: new Date('2021/07/14'),
}
