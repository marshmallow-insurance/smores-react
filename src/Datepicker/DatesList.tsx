import React, { FC } from 'react'
import styled from 'styled-components'
import { getISODay } from 'date-fns'

import { theme } from '../theme'

import { Day } from './types'

type Props = {
  items: Day[]
  handleDateSelect: (date: Date) => void
}

const getBlankDaysCount = (firstDayOfTheMonth: Date) => {
  const dayOfTheWeek = getISODay(firstDayOfTheMonth)
  return dayOfTheWeek - 1
}

export const DatesList: FC<Props> = ({ items, handleDateSelect }) => {
  return (
    <Container>
      {Array(getBlankDaysCount(items[0].date))
        .fill(null)
        .map((_, index) => (
          <ListItem key={`blankItem-${index}`} disabled />
        ))}
      {items.map((item: Day, i) => (
        <ListItem
          key={i}
          disabled={item.disabled}
          className={`ListItem ${item.active ? 'active' : ''}`}
          onClick={() => handleDateSelect(item.date)}
          data-testid={item.label}
        >
          {item.label}
        </ListItem>
      ))}
    </Container>
  )
}

const Container = styled.ul`
  display: grid;
  grid-column-gap: 4px;
  grid-row-gap: 8px;
  grid-template-columns: repeat(7, 1fr);
  padding: 8px;

  @media (min-width: 360px) {
    grid-gap: 8px;
    padding: 16px;
  }

  @media (min-width: 768px) {
    grid-column-gap: 16px;
    padding: 32px;
  }
`

interface IListItem {
  disabled?: boolean
}

const ListItem = styled.li<IListItem>`
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  width: 32px;
  height: 32px;
  border: 1px solid transparent;
  font-weight: ${theme.font.weight.medium};
  font-size: 14px;
  line-height: 15px;
  color: ${theme.colors.secondary};
  background-color: transparent;
  pointer-events: ${({ disabled }) => (disabled ? 'none' : '')};
  opacity: ${({ disabled }) => (disabled ? '0.2' : '1')};
  cursor: pointer;

  &:hover:not(.active) {
    background-color: ${theme.colors.background};
  }

  &.active {
    color: ${theme.colors.white};
    background-color: ${theme.colors.secondary};
    cursor: default;
  }
`
