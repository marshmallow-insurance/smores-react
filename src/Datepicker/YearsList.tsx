import React, { FC } from 'react'
import styled from 'styled-components'

import { theme } from '../theme'

import { focusOutlineStyle } from '../utils/focusOutline'

type Props = {
  handleYearSelect: (year: number) => void
  selectedYear?: number
}

export const YearsList: FC<Props> = ({ handleYearSelect, selectedYear }) => {
  const years = Array.from(Array(200).keys()).map((year) => year + 1900)

  return (
    <Container>
      {years.map((year, i) => (
        <ListButton
          key={i}
          type="button"
          $active={selectedYear === year}
          onClick={() => handleYearSelect(year)}
        >
          {year}
        </ListButton>
      ))}
    </Container>
  )
}

const Container = styled.div`
  justify-items: center;
  display: grid;
  grid-column-gap: 4px;
  grid-row-gap: 8px;
  grid-template-columns: repeat(4, 1fr);
  padding: 8px;
  height: 208px;
  margin-bottom: 16px;
  overflow: scroll;
  width: 304px;

  @media (min-width: 360px) {
    grid-gap: 8px;
    padding: 16px;
  }

  @media (min-width: 768px) {
    grid-column-gap: 16px;
    padding: 32px;
  }
`

interface IListButton {
  disabled?: boolean
  $active?: boolean
}

const ListButton = styled.button<IListButton>`
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 8px;
  width: 50px;
  height: 32px;
  border: 1px solid transparent;
  font-weight: ${theme.font.weight.medium};
  font-size: 14px;
  line-height: 15px;
  color: ${theme.colors.liquorice};
  background-color: transparent;
  pointer-events: ${({ disabled }) => (disabled ? 'none' : '')};
  opacity: ${({ disabled }) => (disabled ? '0.2' : '1')};
  cursor: pointer;

  &:hover:not(.active) {
    color: ${theme.colors.custard};
    background-color: ${theme.colors.liquorice};
  }

  ${({ $active }) =>
    $active &&
    `color: ${theme.colors.custard};
    background-color: ${theme.colors.liquorice};
    cursor: default;`}

  &.active {
    color: ${theme.colors.custard};
    background-color: ${theme.colors.liquorice};
    cursor: default;
  }

  ${focusOutlineStyle}
`
