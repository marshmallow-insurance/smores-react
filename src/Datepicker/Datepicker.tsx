import React, { FC, useState, useLayoutEffect } from 'react'
import styled from 'styled-components'
import {
  format,
  addDays,
  addMonths,
  getMonth,
  getYear,
  isSameDay,
  isSameMonth,
  isWithinInterval,
  startOfMonth,
  getDaysInMonth,
  isToday,
} from 'date-fns'

import { Box } from '../Box'
import { Text } from '../Text'
import { Icon } from '../Icon'
import { theme } from '../theme'

import { DatesList } from './DatesList'
import { Month } from './types'

type DatepickerProps = {
  range?: number
  onDateSelect: (date: string) => void
}

export const Datepicker: FC<DatepickerProps> = ({
  range = 14,
  onDateSelect,
}) => {
  const [activeDay, setActiveDay] = useState<Date>()
  const [activeMonth, setActiveMonth] = useState(0)
  const [listDays, setListDays] = useState<Month[]>([
    {
      monthName: '',
      filteredDays: [],
    },
  ])

  const handleSelectEvent = (date: Date) => {
    setActiveDay(date)
    onDateSelect(format(date, 'yyyy-MM-dd'))
  }

  const generateDaysForMonth = (startDay: Date, endDay: Date) => {
    const month = getMonth(startDay)
    const monthName = format(startDay, 'MMMM')
    const daysInMonth = getDaysInMonth(startDay)
    const year = getYear(startDay)
    const filteredDays = []

    for (let i = 1; i <= daysInMonth; i += 1) {
      const date = new Date(year, month, i)

      filteredDays.push({
        date,
        label: format(date, 'dd'),
        active: activeDay ? isSameDay(date, activeDay) : false,
        disabled:
          !isToday(date) &&
          !isWithinInterval(date, { start: startDay, end: endDay }),
      })
    }

    return {
      monthName,
      filteredDays,
    }
  }

  const combineAvailableDays = () => {
    const today = new Date()
    const endDay = addDays(today, range)
    const sameMonth = isSameMonth(today, endDay)

    let availableDays = []

    // Check if we should generate dates from current for next month
    if (!sameMonth) {
      const thisMonth = generateDaysForMonth(today, endDay)
      const nextMonth = generateDaysForMonth(
        startOfMonth(addMonths(today, 1)),
        endDay,
      )

      availableDays = [thisMonth, nextMonth]
    } else {
      availableDays = [generateDaysForMonth(today, endDay)]
    }

    setListDays(availableDays)
  }

  useLayoutEffect(() => {
    combineAvailableDays()
  }, [activeDay, activeMonth])

  return (
    <Container id="datepicker">
      <Header
        flex
        alignItems="center"
        justifyContent="space-between"
        direction="row"
      >
        <Circle
          disabled={listDays[activeMonth].monthName === listDays[0].monthName}
          onClick={() => setActiveMonth(activeMonth - 1)}
        >
          <Icon render="caret" color="white" size={24} rotate={90} />
        </Circle>

        <Heading tag="h4" typo="desc-base">
          {listDays[activeMonth].monthName}
        </Heading>

        <Circle
          disabled={
            listDays[activeMonth].monthName ===
            listDays[listDays.length - 1].monthName
          }
          onClick={() => setActiveMonth(activeMonth + 1)}
        >
          <Icon render="caret" color="white" size={24} rotate={-90} />
        </Circle>
      </Header>

      <Box flex alignItems="center" justifyContent="center">
        <DatesList
          items={listDays[activeMonth].filteredDays}
          handleDateSelect={handleSelectEvent}
        />
      </Box>
    </Container>
  )
}

const Container = styled(Box)`
  font-family: ${theme.font.system};
  display: inline-block;
  box-sizing: border-box;
  background-color: ${theme.colors.white};
  outline: none;
  border-radius: 8px;
  box-shadow: inset 0 0 0 1px ${theme.colors.grey3};
`

const Header = styled(Box)`
  color: ${theme.colors.blue7};
  user-select: none;
  padding: 16px;
  border-bottom: 1px solid ${theme.colors.grey3};

  span {
    font-size: 16px;
    line-height: 1;
  }
`

const Heading = styled(Text)`
  position: relative;
  top: 2px;
  font-weight: ${theme.font.weight.medium};
`

const Circle = styled.button`
  background-color: ${theme.colors.blue7};
  height: 32px;
  width: 32px;
  border-radius: 50%;
  border: none;
  display: flex;
  justify-content: center;
  padding: 0;
  cursor: pointer;

  :disabled {
    opacity: 0.5;
    cursor: default;
  }
`
