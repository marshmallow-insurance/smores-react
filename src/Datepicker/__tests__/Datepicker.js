import React from 'react'
import { render } from '@testing-library/react'
import 'jest-styled-components'

import { Datepicker } from '../Datepicker'

test('renders', () => {
  const { container } = render(
    <Datepicker range={20} onDateSelect={str => {}} />,
  )
  expect(container.firstChild).toMatchSnapshot()
})
