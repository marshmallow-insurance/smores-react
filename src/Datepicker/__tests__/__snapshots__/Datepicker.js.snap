// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  margin: 0;
  padding: 0;
  color: #112035;
  text-align: left;
  cursor: inherit;
  -webkit-letter-spacing: 0.15px;
  -moz-letter-spacing: 0.15px;
  -ms-letter-spacing: 0.15px;
  letter-spacing: 0.15px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  max-width: 24px;
  height: 24px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  max-width: 24px;
  height: 24px;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.c8 {
  display: grid;
  grid-column-gap: 8px;
  grid-row-gap: 8px;
  grid-template-columns: repeat(7,1fr);
  padding: 16px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4px;
  width: 32px;
  height: 32px;
  border: 1px solid transparent;
  font-weight: 500;
  font-size: 14px;
  line-height: 15px;
  color: #112035;
  background-color: transparent;
  pointer-events: none;
  opacity: 0.2;
  cursor: pointer;
}

.c9:hover:not(.active) {
  background-color: #ECECEC;
}

.c9.active {
  color: #ffffff;
  background-color: #112035;
  cursor: default;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4px;
  width: 32px;
  height: 32px;
  border: 1px solid transparent;
  font-weight: 500;
  font-size: 14px;
  line-height: 15px;
  color: #112035;
  background-color: transparent;
  opacity: 1;
  cursor: pointer;
}

.c10:hover:not(.active) {
  background-color: #ECECEC;
}

.c10.active {
  color: #ffffff;
  background-color: #112035;
  cursor: default;
}

.c0 {
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  display: inline-block;
  box-sizing: border-box;
  background-color: #ffffff;
  outline: none;
  border-radius: 8px;
  box-shadow: inset 0 0 0 1px #E5E5E5;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #112035;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding: 16px;
  border-bottom: 1px solid #E5E5E5;
}

.c1 span {
  font-size: 16px;
  line-height: 1;
}

.c5 {
  position: relative;
  top: 2px;
  font-weight: 500;
}

.c2 {
  background-color: #112035;
  height: 32px;
  width: 32px;
  border-radius: 50%;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
  cursor: pointer;
}

.c2:disabled {
  opacity: 0.5;
  cursor: default;
}

@media (min-width:768px) {
  .c8 {
    grid-column-gap: 16px;
    padding: 32px;
  }
}

<div
  class="c0"
  id="datepicker"
>
  <div
    class="c1"
    direction="row"
  >
    <button
      class="c2"
      disabled=""
    >
      <span
        class="c3"
        color="white"
        rotate="90"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 32 32"
          width="100%"
        >
          <path
            d="M23.66 12.248a1 1 0 0 0-1.412.095L16 19.482l-6.248-7.14a1 1 0 1 0-1.504 1.317l7 8a.995.995 0 0 0 1.504 0l7-8a1 1 0 0 0-.093-1.411z"
            fill="#ffffff"
          />
        </svg>
      </span>
    </button>
    <h4
      class="c4 c5"
      color="blue7"
      cursor="inherit"
    >
      February
    </h4>
    <button
      class="c2"
    >
      <span
        class="c6"
        color="white"
        rotate="-90"
        size="24"
      >
        <svg
          fill="none"
          height="100%"
          viewBox="0 0 32 32"
          width="100%"
        >
          <path
            d="M23.66 12.248a1 1 0 0 0-1.412.095L16 19.482l-6.248-7.14a1 1 0 1 0-1.504 1.317l7 8a.995.995 0 0 0 1.504 0l7-8a1 1 0 0 0-.093-1.411z"
            fill="#ffffff"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    class="c7"
  >
    <ul
      class="c8"
    >
      <li
        class="c9 ListItem "
        disabled=""
      >
        01
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        02
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        03
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        04
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        05
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        06
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        07
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        08
      </li>
      <li
        class="c9 ListItem "
        disabled=""
      >
        09
      </li>
      <li
        class="c10 ListItem "
      >
        10
      </li>
      <li
        class="c10 ListItem "
      >
        11
      </li>
      <li
        class="c10 ListItem "
      >
        12
      </li>
      <li
        class="c10 ListItem "
      >
        13
      </li>
      <li
        class="c10 ListItem "
      >
        14
      </li>
      <li
        class="c10 ListItem "
      >
        15
      </li>
      <li
        class="c10 ListItem "
      >
        16
      </li>
      <li
        class="c10 ListItem "
      >
        17
      </li>
      <li
        class="c10 ListItem "
      >
        18
      </li>
      <li
        class="c10 ListItem "
      >
        19
      </li>
      <li
        class="c10 ListItem "
      >
        20
      </li>
      <li
        class="c10 ListItem "
      >
        21
      </li>
      <li
        class="c10 ListItem "
      >
        22
      </li>
      <li
        class="c10 ListItem "
      >
        23
      </li>
      <li
        class="c10 ListItem "
      >
        24
      </li>
      <li
        class="c10 ListItem "
      >
        25
      </li>
      <li
        class="c10 ListItem "
      >
        26
      </li>
      <li
        class="c10 ListItem "
      >
        27
      </li>
      <li
        class="c10 ListItem "
      >
        28
      </li>
      <li
        class="c10 ListItem "
      >
        29
      </li>
    </ul>
  </div>
</div>
`;
